### PR TITLE
Add tracking to specialist topic tagging page

### PR DESCRIPTION
**Pull request by dnkrj**
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

## What
Adds tracking functionality to the autocomplete using the `onConfirm` callback.
[Uses this functionality to add missing tracking to the specialist topic tagging page.](https://trello.com/c/plXuFwJo/850-add-tracking-to-specialist-topic-tagging-page)

## Why
The current select2 component uses the [`TrackSelectClick` module](https://github.com/alphagov/whitehall/blob/main/app/assets/javascripts/admin_legacy/modules/track_select_click.js) to for GA tracking, this functionality does not work for the accessible autocomplete.
The accessible autocomplete [suggests instead using the `onConfirm` callback](https://github.com/kevindew/accessible-autocomplete#analytics-and-tracking) to perform tracking.

### CODE REVIEW

File: app/assets/javascripts/components/autocomplete.js [hunk 1]
```diff
@@ -23,7 +23,17 @@ window.GOVUK.Modules = window.GOVUK.Modules || {};
...
+      showNoOptionsFound: true,
+      onConfirm: function (value) {
+        var category = $select.getAttribute('data-track-category')
+        var label = $select.getAttribute('data-track-label')
+        if (!category || !label) {
+          return
+        }
+
+        var action = value
+        window.GOVUK.analytics.trackEvent(category, action, { label })
```
Comment (DilwoarH) [inline app/assets/javascripts/components/autocomplete.js:35]: Could we potentially add another attribute called `data-track-select="true"`

```suggestion
        var trackSelect = $select.getAttribute('data-track-select') === "true"
        
        if (trackSelect) {
          var category = $select.getAttribute('data-track-category')
          var label = $select.getAttribute('data-track-label')

          window.GOVUK.analytics.trackEvent(category, value, { label: label })
        }
```

**Review by DilwoarH — CHANGES_REQUESTED** [review]

File: app/assets/javascripts/components/autocomplete.js [hunk 2]
```diff
@@ -23,7 +23,17 @@ window.GOVUK.Modules = window.GOVUK.Modules || {};
...
-      showNoOptionsFound: true
+      showNoOptionsFound: true,
+      onConfirm: function (value) {
+        var category = $select.getAttribute('data-track-category')
+        var label = $select.getAttribute('data-track-label')
+        if (!category || !label) {
+          return
+        }
+
+        var action = value
```
Comment (ollietreend) [inline app/assets/javascripts/components/autocomplete.js:34]: What is the expected value of `action`?

I was hoping to see a unit test which might have given a clue about what values it'd track in different scenarios. But it looks like there isn't a test for this file at all 🙁 and obviously creating one would out-of-scope for this PR.

Looking at the [Accessible Autocomplete docs](https://alphagov.github.io/accessible-autocomplete/#onconfirm-default---), it seems that the `onConfirm` callback will receive an `Object` parameter. But I was expecting `action` to be a `String` for passing in to `GOVUK.analytics.trackEvent`. So based on that, I suspect this tracking might not work as expected 🤔

**Review by ollietreend — COMMENTED** [review]
This approach looks good – just one potential issue in the implementation.

File: app/assets/javascripts/components/autocomplete.js (hunk 2 above)
Comment (dnkrj) [reply app/assets/javascripts/components/autocomplete.js:34]: Yeah, lack of test coverage is really an issue for all of these components.

Action in this case action is a String.
The autocomplete returns [whatever the value of that option was when it was passed into it](https://github.com/kevindew/accessible-autocomplete/blob/master/src/autocomplete.js#L302) when it was initialised.
In this case we're using the enhanceSelectElement function which only [passes the option values to the autocomplete](https://github.com/kevindew/accessible-autocomplete/blob/master/src/wrapper.js#L23) which are the Strings.

File: app/assets/javascripts/components/autocomplete.js (hunk 1 above)
Comment (dnkrj) [reply app/assets/javascripts/components/autocomplete.js:35]: What do you think about this vs. an explicit parameters on the autocomplete component for the tracking category and label?

[the remark this answers is not in the document]
Comment (dnkrj) [reply app/assets/javascripts/components/autocomplete.js:35]: It's the [shorthand for property names](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#property_definitions), which I suppose we shouldn't use if we're still targeting IE at all.

Have updated it.

[the remark this answers is not in the document]
Comment (DilwoarH) [reply app/assets/javascripts/components/autocomplete.js]: nice - didn't know they brought it into JS

**Review by DilwoarH — APPROVED** [review]

**Pull request merged**